File: transformers.py
# coding=UTF-8
from objc import YES, NO, IBAction, IBOutlet
from Foundation import *
from AppKit import *
import re
from fuzzy_dict import FuzzyDict

_NSDateFromString_units = FuzzyDict({
    'second':1,
    'minute':60,
    'hour':60*60,
    'day':60*60*24,
}, cutoff=0.1)

class DidNotUnderstandException:
    def __repr__(self):
        return "I'm sorry, I didn't understand that"

def NSDateFromString(str):
    if str.startswith('in '):
        str = str.strip()
        m = re.match(r'in (?P<num>\d+)\s*(?P<unit>.*)', str)
        if m:
            num, unit = float(m.groupdict()['num']), m.groupdict()['unit']
            global _NSDateFromString_units
            from datetime import datetime, timedelta
            from waker import datetimeToNSDate
            return datetimeToNSDate(datetime.now()+timedelta(seconds=num*_NSDateFromString_units[unit]))
        raise DidNotUnderstandException()
    date = NSDate.dateWithNaturalLanguageString_(str)
    # now convert it to GMT for the internal processing
    from waker import NSDateToDatetime, datetimeToNSDate
    tmp = NSDateToDatetime(date)
    if tmp is None:
        raise DidNotUnderstandException()
    return datetimeToNSDate(tmp)

#2011-01-08 09:00:00 +0100
def test():
    print '--- tests ---'
    import datetime
    from waker import NSTimeZone_tzinfo, datetimeToNSDate, NSDateToDatetime
    # sanity check 
    test_first_jan_2010_8_oclock_datetime = datetime.datetime(2011, 1, 1, 8, 0, tzinfo=NSTimeZone_tzinfo())
    assert(str(test_first_jan_2010_8_oclock_datetime).endswith('08:00:00+00:00'))
    
    # test datetimeToNSDate
    assert(str(datetimeToNSDate(test_first_jan_2010_8_oclock_datetime)).endswith('08:00:00 +0000'))
    
    # test NSDateToDatetime
    nowNSDate = NSDate.date()
    foo = str(nowNSDate)
    bar = str(NSDateToDatetime(nowNSDate))
    #print 'foo', foo
    #print 'bar', bar
    assert(foo[:-6] == bar[:-13])

    #print test_first_jan_2010_8_oclock_datetime
    #print datetimeToNSDate(test_first_jan_2010_8_oclock_datetime)
    print NSDateToDatetime(datetimeToNSDate(test_first_jan_2010_8_oclock_datetime))
    assert(str(NSDateToDatetime(datetimeToNSDate(test_first_jan_2010_8_oclock_datetime))).endswith('08:00:00+00:00'))

    # test NSDateFromString (this uses both the functions above)
    test_first_jan_2010_8_oclock = NSDateFromString('8:00')
    assert(str(test_first_jan_2010_8_oclock).endswith('08:00:00 +0000'))
    
    print '------------'
#test()

class DateTransformer(NSValueTransformer): 
    def transformedValueClass(self):
        return NSString
    transformedValueClass = classmethod(transformedValueClass)
    
    def allowsReverseTransformation(self):
        return YES
    allowsReverseTransformation = classmethod(allowsReverseTransformation)
    
    def transformedValue_(self, value):
        # string in, string out
        if value is None:
            return 'n/a'
        return value
        
    def reverseTransformedValue_(self, value):
        # string in, string out
        if value == 'n/a':
            return None
        return value

class NumberTransformer(NSValueTransformer):
    def transformedValueClass(self):
        return NSNumber
    transformedValueClass = classmethod(transformedValueClass)
    
    def allowsReverseTransformation(self):
        return YES
    allowsReverseTransformation = classmethod(allowsReverseTransformation)
    
    def transformedValue_(self, value):
        return str(value)
        
    def reverseTransformedValue_(self, value):
        if value is None:
            return 0
        return int(value)

class NothingTransformer(NSValueTransformer):
    def transformedValueClass(self):
        return NSString
    transformedValueClass = classmethod(transformedValueClass)
    
    def allowsReverseTransformation(self):
        return YES
    allowsReverseTransformation = classmethod(allowsReverseTransformation)
    
    def transformedValue_(self, value):
        return value
        
    def reverseTransformedValue_(self, value):
        return value
        
def IntToNSColor(value):
    if value is None:
        return NSColor.grayColor()
    return NSColor.colorWithCalibratedRed_green_blue_alpha_(
        ((value & 0xFF0000) >> 16)/255.0, 
        ((value & 0xFF00) >> 8)/255.0, 
        (value & 0xFF)/255.0, 
        1)

class ColorTransformer(NSValueTransformer):
    def transformedValueClass(self):
        return NSColor
    transformedValueClass = classmethod(transformedValueClass)
    
    def allowsReverseTransformation(self):
        return YES
    allowsReverseTransformation = classmethod(allowsReverseTransformation)
    
    def transformedValue_(self, value):
        return IntToNSColor(value)
        
    def reverseTransformedValue_(self, value):
        return int(value.redComponent()*255) << 16 | int(value.greenComponent()*255) << 8 | int(value.blueComponent()*255)

class PredicateTransformer(NSValueTransformer):
    def transformedValueClass(self):
        return NSString
    transformedValueClass = classmethod(transformedValueClass)
    
    def allowsReverseTransformation(self):
        return YES
    allowsReverseTransformation = classmethod(allowsReverseTransformation)
    
    def transformedValue_(self, value):
        if value is None:
            value = 'title CONTAINS ""'
        predicate = NSPredicate.predicateWithFormat_(value)
        if type(predicate) != NSCompoundPredicate:
            predicate = NSCompoundPredicate.andPredicateWithSubpredicates_([predicate])
        return predicate
        
    def reverseTransformedValue_(self, value):
        if value == None:
            return None
        return value.predicateFormat()

class DisableAlarmTransformer(NSValueTransformer):
    def transformedValueClass(self):
        return NSString
    transformedValueClass = classmethod(transformedValueClass)
    
    def allowsReverseTransformation(self):
        return YES
    allowsReverseTransformation = classmethod(allowsReverseTransformation)
    
    def transformedValue_(self, value):
        print 'disable1', value, type(value)
        # NSDate in, string out
        if value is None:
            return True
        return False
        
    def reverseTransformedValue_(self, value):
        print 'disable2', value, type(value)
        if value == False:
            return '8:00'
        return None

class EnableAlarmTransformer(NSValueTransformer):
    def transformedValueClass(self):
        return NSString
    transformedValueClass = classmethod(transformedValueClass)
    
    def allowsReverseTransformation(self):
        return NO
    allowsReverseTransformation = classmethod(allowsReverseTransformation)
    
    def transformedValue_(self, value):
        # NSDate in, string out
        if value is None:
            return False
        return True
        
    def reverseTransformedValue_(self, value):
        if value == True:
            return NSDateFromString('8:00')
        return None

# Create and register value transformers
transformer = DateTransformer.alloc().init()
NSValueTransformer.setValueTransformer_forName_(transformer, u'DateTransformer')

File: transformers.m
 #import "transformers.h"

NSDate* NSDateFromString(NSString* str) {
    if ([str hasPrefix:@"in "]) {
        str = [str stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
        NSError *error = nil;
        NSRegularExpression* re = [NSRegularExpression regularExpressionWithPattern:@"in (\\d+)\\s*(.*)" options:NSRegularExpressionCaseInsensitive error:&error];
        NSTextCheckingResult* m = [re firstMatchInString:str options:0 range:NSMakeRange(0, str.length)];
        if (m) {
            NSRange numRange = [m rangeAtIndex:1];
            NSRange unitRange = [m rangeAtIndex:2];
            float num = [[str substringWithRange:numRange] floatValue];
            NSString* unit = [str substringWithRange:unitRange];
            NSDictionary* lookup = @{
                @"second": @1,
                @"seconds": @1,
                @"secs": @1,
                @"s": @1,
                @"minute": @60,
                @"min": @60,
                @"mins": @60,
                @"m": @60,
                @"hours": @(60*60),
                @"hour": @(60*60),
                @"h": @(60*60),
                @"days": @(60*60*24),
                @"day": @(60*60*24),
                @"d": @(60*60*24),
                };
            if (lookup[unit] != nil) {
                return [[NSDate date] dateWithOffsetDays:0 hours:0 minutes:0 seconds:num * [lookup[unit] intValue]];
            }
        }
//        [[DidNotUnderstandException alloc] raise];
    }
    NSDate* date = [NSDate dateWithNaturalLanguageString:str];
    // TODO: now convert it to GMT for the internal processing
    if (date == nil) {
        [[DidNotUnderstandException alloc] raise];
    }
    return date;
}

NSColor* IntToNSColor(int value) {
    float
        red = ((value & 0xFF0000) >> 16) / 255.0,
        green = ((value & 0xFF00) >> 8) / 255.0,
        blue = (value & 0xFF) / 255.0;
    return [NSColor colorWithCalibratedRed:red green:green blue:blue alpha:1];
}

NSInteger NSColorToInt(NSColor* color) {
    CGFloat red = 0, green = 0, blue = 0, alpha = 0;
    [color getRed:&red green:&green blue:&blue alpha:&alpha];
    int foo = (int)(red * 255) << 16 | (int)(green * 255) << 8 | (int)(blue * 255);
    return foo;
}

@implementation DidNotUnderstandException

- (id)description {
    return @"I'm sorry, I didn't understand that";
}

@end


@implementation DateTransformer

+ (Class)transformedValueClass {
    return [NSString class];
}

+ (BOOL)allowsReverseTransformation {
    return YES;
}

- (id)transformedValue:(id)value {
    // string in, string out
    if (value == nil) {
        return @"n/a";
    }
    return value;
}

- (id)reverseTransformedValue:(id)value {
    // string in, string out
    if ([@"n/a" isEqualToString:value]) {
        return nil;
    }
    return value;
}

@end


@implementation NumberTransformer

+ (Class)transformedValueClass {
    return [NSNumber class];
}

+ (BOOL)allowsReverseTransformation {
    return YES;
}

- (id)transformedValue:(id)value {
    return [value description];
}

- (id)reverseTransformedValue:(id)value {
    if (value == nil) {
        return 0;
    }
    return [NSNumber numberWithInt:[value intValue]];
}

@end


@implementation NothingTransformer

+ (id)transformedValueClass {
    return [NSString class];
}

+ (BOOL)allowsReverseTransformation {
    return YES;
}

- (id)transformedValue:(id)value {
    return value;
}

- (id)reverseTransformedValue:(id)value {
    return value;
}

@end

@implementation ColorTransformer

+ (Class)transformedValueClass {
    return [NSColor class];
}

+ (BOOL)allowsReverseTransformation {
    return YES;
}

- (id)transformedValue:(id)value {
    return IntToNSColor([value intValue]);
}

- (id)reverseTransformedValue:(id)value {
    CGFloat red = 0, green = 0, blue = 0, alpha = 0;
    [value getRed:&red green:&green blue:&blue alpha:&alpha];
    int foo = (int)(red * 255) << 16 | (int)(green * 255) << 8 | (int)(blue * 255);
    return [NSNumber numberWithInt:foo];
}

@end


@implementation PredicateTransformer : NSValueTransformer

+ (Class)transformedValueClass {
    return [NSString class];
}

+ (BOOL)allowsReverseTransformation {
    return YES;
}

- (id)transformedValue:(id)value {
    if (value == nil) {
        value = @"title CONTAINS \"\"";
    }
    NSPredicate* predicate = [NSPredicate predicateWithFormat:value];
    if ([predicate class] != [NSCompoundPredicate class]) {
        predicate = [NSCompoundPredicate andPredicateWithSubpredicates:@[predicate]];
    }
    return predicate;
}

- (id)reverseTransformedValue:(id)value {
    if (value == nil) {
        return nil;
    }
    return [value predicateFormat];
}

@end


@implementation DisableAlarmTransformer

+ (Class)transformedValueClass {
    return [NSString class];
}

+ (BOOL)allowsReverseTransformation {
    return YES;
}

- (id)transformedValue:(id)value {
    //print 'disable1', value, type(value)
    // NSDate in, string out
    if (value == nil) {
        return @YES;
    }
    return @NO;
}

- (id)reverseTransformedValue:(id)value {
    //print 'disable2', value, type(value)
    if (value == nil) {
        return @"8:00";
    }
    return nil;
}

@end


@implementation EnableAlarmTransformer

+ (Class)transformedValueClass {
    return [NSString class];
}

+ (BOOL)allowsReverseTransformation {
    return NO;
}

- (id)transformedValue:(id)value {
    // NSDate in, string out
    if ([value boolValue] == NO) {
        return @NO;
    }
    return @YES;
}

- (id)reverseTransformedValue:(id)value {
    if ([value boolValue] == YES) {
        return NSDateFromString(@"8:00");
    }
    return nil;
}

@end
